src/Stack/List.hs

```diff
@@ -0,0 +1,77 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+module Stack.List
+  ( listPackages
+  ) where
+
+import Stack.Prelude
+import qualified RIO.Map as Map
+import RIO.List (intercalate)
+import RIO.Process (HasProcessContext)
+
+newtype ListException
+  = CouldNotParsePackageSelectors [String]
+    deriving Typeable
+instance Exception ListException
+instance Show ListException where
+    show (CouldNotParsePackageSelectors strs) = unlines $ map ("- " ++) strs
+
+-- | Intended to work for the command line command.
+listPackages
+  :: forall env. (HasPantryConfig env, HasLogFunc env, HasProcessContext env)
+  => Maybe RawSnapshot -- ^ when looking up by name, take from this build plan
+  -> [String] -- ^ names or identifiers
+  -> RIO env ()
+listPackages mSnapshot input = do
+    let (errs1, names) = case mSnapshot of
+                   Just snapshot | null input ->
+                                     ([], Map.keys (rsPackages snapshot))
+                   _ -> partitionEithers $ map parse input
+    (errs2, locs) <- partitionEithers <$> traverse toLoc names
+    case errs1 ++ errs2 of
+      [] -> pure ()
+      errs -> throwM $ CouldNotParsePackageSelectors errs
+    mapM_ (logInfo . fromString . packageIdentifierString) locs
+  where
+    toLoc | Just snapshot <- mSnapshot = toLocSnapshot snapshot
+          | otherwise = toLocNoSnapshot
+
+    -- FIXME fallback to global or project resolver
```

Comment:
@juhp what do you mean by this FIXME? Do you want to change the way `stack list` works with no resolver specified? Do you think the way it works in this implementation is bad in some sense?

Reply:
Good catch - I was looking at this too earlier, and then forgot to remove it.
For now I feel falling back to the Hackage is probably good enough and useful , since `stack list` is not a project command.
So I think I will just drop the comment (I had left it there as a placeholder/reminder about this decision).